### ISSUE TITLE
DataStore tries to sync before logged in

### ISSUE THREAD

**Issue by raywhiteside** [aws-amplify/amplify-flutter#917]
**Describe the bug**
This exception is thrown when Amplify is configured with DataStore:

```
E/amplify:aws-datastore(14092): DataStoreException{message=Error during subscription., cause=ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
```

**To Reproduce**
Configure Amplify with Cognito, API, and Datastore.

**Expected behavior**
Recently DataStore.start() was introduced. If the developer has the option to determine when DataStore should actually begin syncing, I don't expect to get a an exception immediately after configuration with no chance to execute a login flow first.

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[x] Android
[] iOS

  
  ```flutter doctor -v
[✓] Flutter (Channel stable, 2.5.1, on Arch Linux 5.14.6-arch1-1, locale en_US.UTF-8)
    • Flutter version 2.5.1 at /opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffb2ecea52 (10 days ago), 2021-09-17 15:26:33 -0400
    • Engine revision b3af521a05
    • Dart version 2.14.2

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /opt/android-sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /opt/android-sdk
    • ANDROID_SDK_ROOT = /opt/android_sdk
    • Java binary at: /opt/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at google-chrome

[✓] Linux toolchain - develop for Linux desktop
    • clang version 12.0.1
    • cmake version 3.21.3
    • ninja version 1.10.2
    • pkg-config version 1.8.0

[✓] Android Studio (version 2020.3)
    • Android Studio at /opt/android-studio
    • Flutter plugin version 59.0.2
    • Dart plugin version 203.8292
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] Connected device (3 available)
    • sdk gphone x86 64 (mobile) • emulator-5554 • android-x64    • Android 11 (API 30) (emulator)
    • Linux (desktop)            • linux         • linux-x64      • Arch Linux 5.14.6-arch1-1
    • Chrome (web)               • chrome        • web-javascript • Google Chrome 94.0.4606.61

• No issues found!

```

**Additional context**

Example code:

```
// core packages
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:pedantic/pedantic.dart';

// amplify packages
import 'package:amplify_flutter/amplify.dart';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_api/amplify_api.dart';
import 'package:amplify_datastore/amplify_datastore.dart';

// amplify project code
import 'package:bbp_todo/amplifyconfiguration.dart';
import 'package:bbp_todo/models/ModelProvider.dart';

// global constants
import 'package:bbp_todo/constants/loggers.dart';

class AmplifyTest extends StatefulWidget {
  @override
  _AmplifyTestState createState() => _AmplifyTestState();
}

class _AmplifyTestState extends State<AmplifyTest> {
  @override
  void initState() {
    super.initState();
    _configureAmplify();
  }

  void _configureAmplify() async {

    final AmplifyAuthCognito _cognito = AmplifyAuthCognito();
    final AmplifyAPI _api = AmplifyAPI();
    final AmplifyDataStore _dataStore =
        AmplifyDataStore(modelProvider: ModelProvider.instance);

    await Amplify.addPlugin(_cognito);
    await Amplify.addPlugin(_api);
    await Amplify.addPlugin(_dataStore);
    try {
      await wtf('About to configure amplify');
      await Amplify.configure(amplifyconfig);
      await wtf('Finished configuring amplify');
      await Future.delayed(const Duration(seconds: 10));
      await wtf('Finished waiting');
    } on AmplifyAlreadyConfiguredException {
      unawaited(info('Amplify called configuration again, '
          'normal for hot reload'));
    } catch (e) {
      unawaited(error('Amplify configure error ${e.toString()}'));
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'title',
      theme: ThemeData(
        primarySwatch: Colors.orange,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: Scaffold(
          appBar: AppBar(
              title: const Text('wat'),
          ),
      ),
    );
  }
}
```

Log:

```
Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on sdk gphone x86 64 is available at: http://127.0.0.1:40461/aB8BeorvmNw=/
Activating Dart DevTools...
I/amplify:flutter:api(14092): Added API plugin
I/flutter (14092): ┌───────────────────────────────────────────────────────────────────────────────
I/flutter (14092): │ 👾 About to configure amplify
I/flutter (14092): └───────────────────────────────────────────────────────────────────────────────
D/AWSMobileClient(14092): Using the SignInProviderConfig from `awsconfiguration.json`.
D/AWSMobileClient(14092): Inspecting user state details
Activating Dart DevTools...                                      1,226ms
I/amplify:aws-datastore(14092): Orchestrator lock acquired.
I/flutter (14092): ┌───────────────────────────────────────────────────────────────────────────────
I/flutter (14092): │ 👾 Finished configuring amplify
I/flutter (14092): └───────────────────────────────────────────────────────────────────────────────
I/amplify:aws-datastore(14092): Orchestrator transitioning from STOPPED to SYNC_VIA_API
I/amplify:aws-datastore(14092): Starting to observe local storage changes.
I/amplify:aws-datastore(14092): Now observing local storage. Local changes will be enqueued to mutation outbox.
I/amplify:aws-datastore(14092): Setting currentState to LOCAL_ONLY
I/amplify:aws-datastore(14092): Setting currentState to SYNC_VIA_API
I/amplify:flutter:datastore(14092): Established a new stream form flutter com.amplifyframework.datastore.storage.sqlite.-$$Lambda$0NKeZW6Mwf4F83r11OGUIWtYc5c@bce6b66
I/amplify:aws-datastore(14092): Orchestrator lock released.
I/amplify:aws-datastore(14092): Starting API synchronization mode.
I/amplify:aws-datastore(14092): Starting processing subscription events.
E/amplify:aws-datastore(14092): Failure encountered while attempting to start API sync.
E/amplify:aws-datastore(14092): DataStoreException{message=Error during subscription., cause=ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
E/amplify:aws-datastore(14092): 	at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:329)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$8gXTT9bgLZGP51blif6hkh7EwIc.accept(Unknown Source:4)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionOperation.lambda$null$1$SubscriptionOperation(SubscriptionOperation.java:87)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$EfR-sAOj2xzA_-vAoxWBdb2iFk8.accept(Unknown Source:4)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:131)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:77)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$Uz-v98-sFPcWrrCLK25EF8LKbXo.run(Unknown Source:2)
E/amplify:aws-datastore(14092): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
E/amplify:aws-datastore(14092): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/amplify:aws-datastore(14092): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/amplify:aws-datastore(14092): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/amplify:aws-datastore(14092): 	at java.lang.Thread.run(Thread.java:923)
E/amplify:aws-datastore(14092): Caused by: ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.fetchToken(DefaultCognitoUserPoolsAuthProvider.java:98)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.getLatestAuthToken(DefaultCognitoUserPoolsAuthProvider.java:104)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.forCognitoUserPools(SubscriptionAuthorizer.java:147)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.createHeaders(SubscriptionAuthorizer.java:96)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.createHeadersForConnection(SubscriptionAuthorizer.java:71)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.buildConnectionRequestUrl(SubscriptionEndpoint.java:300)
E/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:127)
E/amplify:aws-datastore(14092): 	... 7 more
W/amplify:aws-datastore(14092): API sync failed - transitioning to LOCAL_ONLY.
W/amplify:aws-datastore(14092): DataStoreException{message=Error during subscription., cause=ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Evaluate details.}
W/amplify:aws-datastore(14092): 	at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$subscription$3(AppSyncClient.java:329)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$8gXTT9bgLZGP51blif6hkh7EwIc.accept(Unknown Source:4)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionOperation.lambda$null$1$SubscriptionOperation(SubscriptionOperation.java:87)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$EfR-sAOj2xzA_-vAoxWBdb2iFk8.accept(Unknown Source:4)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:131)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionOperation.lambda$start$2$SubscriptionOperation(SubscriptionOperation.java:77)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.-$$Lambda$SubscriptionOperation$Uz-v98-sFPcWrrCLK25EF8LKbXo.run(Unknown Source:2)
W/amplify:aws-datastore(14092): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
W/amplify:aws-datastore(14092): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
W/amplify:aws-datastore(14092): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
W/amplify:aws-datastore(14092): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
W/amplify:aws-datastore(14092): 	at java.lang.Thread.run(Thread.java:923)
W/amplify:aws-datastore(14092): Caused by: ApiAuthException{message=You must be signed-in with Cognito Userpools to be able to use getTokens, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.fetchToken(DefaultCognitoUserPoolsAuthProvider.java:98)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.getLatestAuthToken(DefaultCognitoUserPoolsAuthProvider.java:104)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.forCognitoUserPools(SubscriptionAuthorizer.java:147)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.createHeaders(SubscriptionAuthorizer.java:96)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionAuthorizer.createHeadersForConnection(SubscriptionAuthorizer.java:71)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.buildConnectionRequestUrl(SubscriptionEndpoint.java:300)
W/amplify:aws-datastore(14092): 	at com.amplifyframework.api.aws.SubscriptionEndpoint.requestSubscription(SubscriptionEndpoint.java:127)
W/amplify:aws-datastore(14092): 	... 7 more
I/amplify:aws-datastore(14092): Orchestrator transitioning from SYNC_VIA_API to LOCAL_ONLY
I/amplify:aws-datastore(14092): Setting currentState to LOCAL_ONLY
I/amplify:aws-datastore(14092): Stopping subscription processor.
I/amplify:aws-datastore(14092): Stopped subscription processor.
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
W/CognitoUserSession(14092): CognitoUserSession is not valid because idToken is null.
W/AWSMobileClient(14092): signalTokensNotAvailable
W/InternalCallback(14092): Library attempted to call user callback twice, expected only once
I/flutter (14092): ┌───────────────────────────────────────────────────────────────────────────────
I/flutter (14092): │ 👾 Finished waiting
I/flutter (14092): └───────────────────────────────────────────────────────────────────────────────
```

**Comment by HuiSF** [comment]
Hi @raywhiteside thanks for opening this issue.

This behavior is expected while using amplify-ios and amplify-android (the dependencies of amplify-flutter). An initial sync is expected right after plugin configuration is completed regardless signing in state, which triggered the error in your logs. When this exception happens it falls back to local only mode. In addition, normally, both iOS and Android try to resume sync [on every DataStore operation (CRUD) APIs calls](https://docs.amplify.aws/lib/datastore/other-methods/q/platform/flutter/#start). 

The `start()` API resumes API sync without invoke any operation APIs. A useful use case is you can invoke this API right after a user signs in successfully to resume the API sync.

**Comment by HuiSF** [comment]
Hi @raywhiteside a follow up, by any chance, did you invoke `Amplify.DataStore.observe()` API right after completing the plugin configuration?

**Comment by raywhiteside** [comment]
> Hi @raywhiteside a follow up, by any chance, did you invoke Amplify.DataStore.observe() API right after completing the plugin configuration?

No, the code I pasted is everything I am running. I'm simply wrapping the above widget in `runApp()` in `main.dart`. 

> An initial sync is expected right after plugin configuration is completed regardless signing in state, which triggered the error in your logs. When this exception happens it falls back to local only mode. 

If this is expected, can't this exception be caught? Even better, why can't it check to see if it has the necessary credentials before attempting to sync?

I get that this is technically innocuous, but it's a pretty bold exception to be throwing on the happy-path. This muddies debug for other issues, including having to filter this error out when doing automated checking of regression logs for failures. I was waiting on pushing this issue because I thought `start()` would allow me to resolve it, but if that's not the case I think this needs to be handled better within Amplify.

**Comment by Jordan-Nelson** [comment]
I believe that DataStore starts the sync process immediately because amplify-flutter is invoking `observe()` from amplify-ios/amplify-android as part of the configuration for datastore. The config method invokes `setUpObserve` on the method channel [here](https://github.com/aws-amplify/amplify-flutter/blob/main/packages/amplify_datastore/lib/method_channel_datastore.dart#L90), which invokes observe on the native plugin (see [example for Android](https://github.com/aws-amplify/amplify-flutter/blob/aefce2c8edfac287ade8897856081adb24d7c68e/packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt#L382)).

In stead of setting up the observe subscription immediately, it could be set up after the first time a DataStore operation (`query()`, `save()`, `delete()`, `observe()`) is run, or until `start()` is explicitly called. 

@HuiSF What do you think?

**Comment by HuiSF** [comment]
Thanks for the details @raywhiteside I agree this is no a good developing experience.
And I agree with @Jordan-Nelson we probably need to avoid invoke native Android `observe()` until an explicit instruction made to do so.

**Comment by HuiSF** [comment]
Closing as duplicate of #736 .
